archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,371 @@
+  (hM : M = {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k }) :
+  card M ≤ (2 ^ k) * nat.sqrt(x) :=
+begin
+  set f : ℕ × ℕ → ℕ := λ mn, (mn.2 + 1) ^ 2 * (mn.1 + 1) - 1 with hf_def,
+  set M₁ := {e ∈ range x | squarefree (e + 1) ∧ ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k}
+            with hM₁,
+  set M₂ := {e ∈ range (nat.sqrt x) | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k} with hM₂,
+  set K := finset.product M₁ M₂ with hK,
+
+  have h1 : M ⊆ finset.image f K,
+  { intros m hm,
+    simp only [hM, finset.sep_def, finset.mem_filter, finset.mem_range] at hm,
+    obtain ⟨hmx, hmp⟩ := hm,
+    have h1' : ∃ (a b : ℕ), (a ∈ M₁ ∧ b ∈ M₂) ∧ f (a, b) = m,
+    { obtain ⟨a, b, hab₁, hab₂⟩ := sq_mul_squarefree' (zero_lt_succ m),
+
+      have h11 : a ∈ M₁,
+      { simp only [hM₁, finset.sep_def, finset.mem_filter, finset.mem_range],
+        have ham : (a + 1) ∣ (m + 1) := dvd.intro_left ((b + 1) ^ 2) hab₁,
+        refine ⟨(lt_of_le_of_lt _) hmx, hab₂, _⟩,
+        { exact succ_le_succ_iff.mp ((nat.le_of_dvd (zero_lt_succ m)) ham) },
+        { intro p,
+          specialize hmp p,
+          rintros ⟨hpp, hpa⟩,
+          exact hmp ⟨hpp, dvd.trans hpa ham⟩ } },
+
+      have h12 : b ∈ M₂,
+      { simp only [hM₂, finset.sep_def, finset.mem_filter, finset.mem_range],
+        have hbm₁ : (b + 1) ^ 2 ∣ (m + 1) := dvd.intro (a + 1) hab₁,
+        have hbm₂ : (b + 1) ∣ (m + 1) := nat.dvd_of_pow_dvd one_le_two hbm₁,
+        split,
+        { calc b < b + 1 : lt_add_one b
+          ...    ≤ nat.sqrt(m + 1) : by { rw [nat.le_sqrt, ← pow_two (b + 1)],
+                                          exact nat.le_of_dvd (zero_lt_succ m) hbm₁ }
+          ...    ≤ nat.sqrt x : sqrt_le_sqrt (succ_le_iff.mpr hmx) },
+        { intro p,
+          specialize hmp p,
+          rintros ⟨hpp, hpb⟩,
+          exact hmp ⟨hpp, dvd.trans hpb hbm₂⟩ } },
+
+      have h13 : f (a, b) = m,
+      { have h13' : (b + 1) ^ 2 * (a + 1) - 1 = m, rw [hab₁, succ_sub_one m],
+        simpa [hf_def] },
+
+      exact ⟨a, b, ⟨h11, h12⟩, h13⟩ },
+    simpa },
+
+  have h2 : card M₁ ≤ 2 ^ k := lemma4_aux_card_M1_le_2_pow_k hM₁,
+  have h3 : card M₂ ≤ nat.sqrt x,
+  { rw ← card_range (nat.sqrt x), refine card_le_of_subset _, simp [hM₂] },
+
+  calc card M ≤ card (finset.image f K) : finset.card_le_of_subset h1
+  ...         ≤ card K                  : finset.card_image_le
+  ...         = card M₁ * card M₂       : finset.card_product M₁ M₂
+  ...         ≤ 2 ^ k * nat.sqrt x      : mul_le_mul' h2 h3,
+end
+
+theorem real.tendsto_sum_one_div_prime_at_top :
+  filter.tendsto (λ n, ∑ p in { p ∈ range n | nat.prime p }, (1 / (p : ℝ))) at_top at_top :=
+begin
+  by_contradiction,
+
+  obtain ⟨k, h1⟩ := lemma1_not_hyp_imp_sum_lt_half h,
+  set x := (2 ^ (k + 1)) * (2 ^ (k + 1)) with hxk,
+  set P := {p ∈ range (x + 1) | p > k ∧ nat.prime p } with hP,
+  specialize h1 x,
+  rw ← hP at h1,
+
+  set M := {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k } with hM,
```

Comment:
Why do you use this style instead of just `let M := `? It seems you could avoid the equality arguments to the lemmas.

Reply:
If I use `let M := ?` in the main theorem, and remove the equality arguments in the lemmas, then the rewrites don't work. For example, if I change lemma2 to this:
```
lemma lemma2_range_x_sdiff_M_eq_U (x k : ℕ) : 
  finset.range x \ {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k }
  = finset.bUnion {p ∈ range (x + 1) | p > k ∧ nat.prime p } (λ p, {e ∈ range x | p ∣ (e + 1) }) :=
```
Then the rw ← lemma2_range_x_sdiff_M_eq_U x k in the main theorem doesn't work.
```
  let M := {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k },
  let N := (λ p, {e ∈ range x | p ∣ (e + 1) }),
  let U := finset.bUnion {p ∈ range (x + 1) | p > k ∧ nat.prime p } N,

  have h2 : x = card U + card M,
  { rw [← card_range x, ← lemma2_range_x_sdiff_M_eq_U x k],
```

Is there a different way of doing this that would work?

Reply:
I'm also interested in this; I often use `let` but have to pepper my rewrites with `dsimp only [f]` or whatever